lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,378 @@
+# frozen_string_literal: true
+
+module RGeo
+  module Cartesian
+    # A Doubly Connected Edge List (DCEL) implementation of a Planar Graph.
+    # It represents geometries as vertices and half-edges.
+    #
+    # It includes an incident_edges hash that maps vertices to an array
+    # of half-edges whose origins are that vertex.
+    #
+    # Upon instantiation, the graph will compute the intersections using
+    # the SweeplineIntersector, populate the incident_edges map, and
+    # link all cyclic edges.
+    class PlanarGraph
+      # HalfEdge represents an edge as 2 directed edges.
+      # One half-edge will have it's origin at edge.s, the other
+      # at edge.e. Both half-edges will be linked as each other's twins.
+      #
+      # HalfEdges also contain references to the next and prev half-edges,
+      # where next's origin is this half-edges destination. Prev's destination
+      # is this half-edge's origin.
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          e1 = new(edge.s)
+          e2 = new(edge.e)
+
+          e1.twin = e2
+          e2.twin = e1
+          [e1, e2]
+        end
+
+        def initialize(origin)
+          @origin = origin
+          @twin = nil
+          @next = nil
+          @prev = nil
+        end
+        attr_reader :origin
+        attr_accessor :twin, :next, :prev
+
+        # HalfEdges will be sorted around their shared vertex
+        # in a CW fashion. This means that face interiors will be
+        # a CCW.
+        def <=>(other)
+          angle <=> other.angle
+        end
+
+        # Will find attempt to find a cycle starting at this
+        # HalfEdge. Will end upon finding the first repeated HalfEdge
+        # or a HalfEdge where +next+ is nil.
+        #
+        # If a block is given, each HalfEdge seen will be yielded to the block.
+        #
+        # @return [Set]
+        def and_connected
+          hedges = Set.new
+          yield(self) if block_given?
+          hedges << self
+
+          n = self.next
+          until hedges.include?(n) || n.nil?
+            yield(n) if block_given?
+            hedges << n
+            n = n.next
+          end
+          hedges
+        end
```

Comment:
```suggestion
        # If a block is given, each HalfEdge seen will be yielded to the block, otherwise
        # returns an enumerator
        def and_connected
          return to_enum(__method__) unles block_given?

          hedges = Set.new
          yield(self)
          hedges << self

          n = self.next
          until hedges.include?(n) || n.nil?
            yield(n)
            hedges << n
            n = n.next
          end
        end
```

I guess you could always return an enum!

Reply:
Good idea! Done.